### PR TITLE
[ExportVerilog] Use pretty-printer for improved output formatting.

### LINKED ISSUE #4181: [ExportVerilog] Expressions substituted into verbatim expressions should not wrap
Current wrapping behavior wraps expressions if they're longer than the configured `emittedLineLength`.

This can result in verbatim statements containing newlines and indentation which does not seem in spirit of verbatim.

Test case:

```mlir
hw.module @VerbatimWrapping(%clock : i1, %cond : i1, %val : i8, %a : i3, %b : i3) {
      %x = comb.add %a, %b : i3
      %y = comb.xor %a, %b : i3
      %arr = hw.array_create %x, %y, %x, %y, %x, %y, %x, %y, %x, %y, %x, %y : i3
      sv.verbatim "// {{0}}" (%arr) : !hw.array<12xi3>
}
```

This verbatim attempts to prefix an expression with `//` to make it a comment.

Currently this is the output of `circt-opt -export-verilog -o /dev/null`:
```systemverilog
// Generated by CIRCT 1.20.0g20221025_97824a4
module VerbatimWrapping(	// ./verbatim-wrapping.mlir:1:1
  input       clock,
              cond,
  input [7:0] val,
  input [2:0] a,
              b);

  wire [2:0] _GEN = a + b;	// ./verbatim-wrapping.mlir:2:12
  wire [2:0] _GEN_0 = a ^ b;	// ./verbatim-wrapping.mlir:3:12
  // {{_GEN}, {_GEN_0}, {_GEN}, {_GEN_0}, {_GEN}, {_GEN_0}, {_GEN}, {_GEN_0}, {_GEN}, {_GEN_0},
                {_GEN}, {_GEN_0}}	// ./verbatim-wrapping.mlir:2:12, :3:12, :4:14, :5:7
endmodule
```

Which is not valid verilog due to the unexpected wrapping occurring.

### LINKED ISSUE #4182: [ExportVerilog] When "commenting-out" zero width contents, wrapping may occur and break the output
In a number of places when emitting a construct that's zero-width instead of omitting it we do so but prefix it with `//`.  This is fragile as the emitted text following can often trigger wrapping and cause the output to be invalid.

Example (modified from `hw-dialect.mlir`), `zerowidth-wrapping.mlir`:
```mlir
hw.module @TestZero(%a: i4, %zeroBitWithAVeryLongNameWhichMightSeemUnlikelyButHappensAllTheTime: i0, %arrZero: !hw.array<3xi0>)
  -> (r0: i4, rZeroOutputWithAVeryLongName_YepThisToo_LongNamesAreTheWay_MoreText_GoGoGoGoGo: i0, arrZero_0: !hw.array<3xi0>) {

  %b = comb.add %a, %a : i4
  %c = comb.add %zeroBitWithAVeryLongNameWhichMightSeemUnlikelyButHappensAllTheTime, %zeroBitWithAVeryLongNameWhichMightSeemUnlikelyButHappensAllTheTime : i0
  hw.output %b, %c, %arrZero : i4, i0, !hw.array<3xi0>
}

hw.module @TestZeroInstance(%aa: i4, %azeroBit: i0, %aarrZero: !hw.array<3xi0>)
  -> (r0: i4, rZeroOutputWithAVeryLongNameYepThisToo: i0, arrZero_0: !hw.array<3xi0>) {

  %o1, %o2, %o3 = hw.instance "iii" @TestZero(a: %aa: i4, zeroBitWithAVeryLongNameWhichMightSeemUnlikelyButHappensAllTheTime: %azeroBit: i0, arrZero: %aarrZero: !hw.array<3xi0>) -> (r0: i4, rZeroOutputWithAVeryLongName_YepThisToo_LongNamesAreTheWay_MoreText_GoGoGoGoGo: i0, arrZero_0: !hw.array<3xi0>)
 %c = comb.add %o2, %o2 : i0

  hw.output %o1, %c, %o3 : i4, i0, !hw.array<3xi0>
}
```

Here's the current output of `circt-opt -export-verilog ./zerowidth-wrapping.mlir -o /dev/null`:
```systemverilog
// Generated by CIRCT 1.20.0g20221025_97824a4
module TestZero(	// ./zerowidth-wrapping.mlir:1:1
     input  [3:0]               a,
  // input  /*Zero Width*/      zeroBitWithAVeryLongNameWhichMightSeemUnlikelyButHappensAllTheTime,
  // input  [2:0]/*Zero Width*/ arrZero,
     output [3:0]               r0
  // output /*Zero Width*/      rZeroOutputWithAVeryLongName_YepThisToo_LongNamesAreTheWay_MoreText_GoGoGoGoGo
  // output [2:0]/*Zero Width*/ arrZero_0
   );

  assign r0 = a + a;	// ./zerowidth-wrapping.mlir:4:8, :6:3
  // Zero width: assign rZeroOutputWithAVeryLongName_YepThisToo_LongNamesAreTheWay_MoreText_GoGoGoGoGo = zeroBitWithAVeryLongNameWhichMightSeemUnlikelyButHappensAllTheTime +
                zeroBitWithAVeryLongNameWhichMightSeemUnlikelyButHappensAllTheTime;	// ./zerowidth-wrapping.mlir:5:8, :6:3
  // Zero width: assign arrZero_0 = arrZero;	// ./zerowidth-wrapping.mlir:6:3
endmodule

module TestZeroInstance(	// ./zerowidth-wrapping.mlir:9:1
     input  [3:0]               aa,
  // input  /*Zero Width*/      azeroBit,
  // input  [2:0]/*Zero Width*/ aarrZero,
     output [3:0]               r0
  // output /*Zero Width*/      rZeroOutputWithAVeryLongNameYepThisToo
  // output [2:0]/*Zero Width*/ arrZero_0
   );

  // Zero width: wire /*Zero Width*/ _iii_rZeroOutputWithAVeryLongName_YepThisToo_LongNamesAreTheWay_MoreText_GoGoGoGoGo;	// ./zerowidth-wrapping.mlir:12:19
  TestZero iii (	// ./zerowidth-wrapping.mlir:12:19
    .a                                                                              (aa),
  //.zeroBitWithAVeryLongNameWhichMightSeemUnlikelyButHappensAllTheTime             (azeroBit),
  //.arrZero                                                                        (aarrZero),
    .r0                                                                             (r0)
  //.rZeroOutputWithAVeryLongName_YepThisToo_LongNamesAreTheWay_MoreText_GoGoGoGoGo (_iii_rZeroOutputWithAVeryLongName_YepThisToo_LongNamesAreTheWay_MoreText_GoGoGoGoGo)
  //.arrZero_0                                                                      (arrZero_0)
  );
  // Zero width: assign rZeroOutputWithAVeryLongNameYepThisToo = _iii_rZeroOutputWithAVeryLongName_YepThisToo_LongNamesAreTheWay_MoreText_GoGoGoGoGo +
                _iii_rZeroOutputWithAVeryLongName_YepThisToo_LongNamesAreTheWay_MoreText_GoGoGoGoGo;	// ./zerowidth-wrapping.mlir:12:19, :13:7, :15:3
endmodule
```

With the problematic lines being:

```systemverilog
  // Zero width: assign rZeroOutputWithAVeryLongName_YepThisToo_LongNamesAreTheWay_MoreText_GoGoGoGoGo = zeroBitWithAVeryLongNameWhichMightSeemUnlikelyButHappensAllTheTime +
                zeroBitWithAVeryLongNameWhichMightSeemUnlikelyButHappensAllTheTime;	// ./zerowidth-wrapping.mlir:5:8, :6:3
```

and

```systemverilog
  // Zero width: assign rZeroOutputWithAVeryLongNameYepThisToo = _iii_rZeroOutputWithAVeryLongName_YepThisToo_LongNamesAreTheWay_MoreText_GoGoGoGoGo +
                _iii_rZeroOutputWithAVeryLongName_YepThisToo_LongNamesAreTheWay_MoreText_GoGoGoGoGo;	// ./zerowidth-wrapping.mlir:12:19, :13:7, :15:3
```

I believe the instance port expression can be made to wrap as well, and perhaps others, but haven't pushed on triggering that case.

### PR DESCRIPTION
**Pull request by dtzSiFive**
Builds on top of the pretty-printing code in #4192 .

Produces much more readable in many cases (IMHO) especially with complex expressions and statements containing non-trivial expressions.

Can see before/after differences here: https://dtzsifive.github.io/circt-pp-results/ (pages may change as things develop, output not final, etc.).  Also, [chipyard](https://gist.githubusercontent.com/dtzSiFive/edde310272aff78182b2e3235c2ba6fd/) which has things like `if` statements.

Not everything is possible, but there are a lot of decisions/stylistic choices (to be) made, if you're interested please take a look / give it a go and let me know what you think so we can iterate towards our preferred output.

Testing / feedback on specifics or in general is greatly appreciated!

Fixes #4181.
Fixes #4182.

Next steps:
* [x] make CI happy / fix tests.
* [x] Fix wrapping verbatim / zero-width "commented-out" lines (similar to #4181, #4182, but much easier to encounter).
* [x] Revisit emission of statements marked TODO as they work but were ported quickly and should be tweaked for better output.

* [x] Testing: functional (generated output "should be" equivalent when flattened, but...)
* [x] Testing: performance / memory.  One goal was to not make emission slower (faster would be nice :wink:), as we especially need to scale well on large designs.  In my testing so far things look good, but this has not been extensively or carefully measured yet.
* [x] Testing: linters?

### CODE REVIEW

**Comment by dtzSiFive** [comment]
(rebased after fixing tidy issues in ExportVerilog)

File: test/Conversion/ExportVerilog/pretty.mlir
Comment (dtzSiFive) [reply test/Conversion/ExportVerilog/pretty.mlir:27]: Thanks for the feedback!  Let's iterate towards best output we can!

So braced list emission does this, and on some giant lists it can be ...sub-optimal.  I'd like to improve it especially for lists of elements that cannot break themselves (like example here).  Presently the comma-separated entries are put in a "cbox" which means all breaks within are broken or none are (not including breaks within subgroups).  Token stream is something like:

```
STRING({), BEGIN(CBOX), BREAK(0, INDENT), STRING(_GEN_1), STRING(,), BREAK(1), STRING(_GEN_2), STRING{,}, BREAK(1), ..., BREAK(0, -INDENT), END, STRING(})
```

Which means if the list doesn't fit, the group breaks together.   This is done to avoid output like:

```
a, b, c,
b + c + d
+ e, a + b
+ c
```

Which IMO is undesirable and hard to read.

---

Implementation-wise I'm not sure how best to create better output with the tools we have, but I'll think some more.
I tried making it so if one expression breaks the break after its comma is forced, but at the time couldn't make that work.

A few less-than-ideal ideas:

1. we further lean on the fact that we buffer tokens entirely when emitting any expression (for paren/cast insertion later) to scan the emitted tokens for breaks and switch to an "ibox" in that case.  Oof.

2. If the common-case simple expressions can be checked with a quick walk (maybe a short whitelist of operation types?) we could emit an "ibox" instead of the cbox (and no breaks next to braces) which does the "pack until line is full, break, continue to pack that line" behavior I think you're saying you prefer (if I I understand correctly).

FWIW both of these are operating under "let's pack these only if they're simple" idea, which maybe isn't best, LMK what behavior might be better!

There's likely other approaches/ideas, but does the above make sense? LMK what you think-- either in terms of preferred output or how to do it!

Comment (dtzSiFive) [reply test/Conversion/ExportVerilog/pretty.mlir:27]: Actually, what output are you saying you prefer? Can you show a short example (whatever demonstrates it, not necessarily one from the lit test)?  Thanks!

Comment (uenoku) [reply test/Conversion/ExportVerilog/pretty.mlir:27]: I was wondering we can do something like this (in the future):
```
a, b, c, 
b + c + d 
+ e, 
a + b + c
```

Instead of inserting new lines to every operand:
```
a, 
b, 
c, 
b + c + d 
+ e, 
a + b + c
```

Comment (dtzSiFive) [reply test/Conversion/ExportVerilog/pretty.mlir:27]: Great, that's my preferred output too.  I'll think about how to do this.

**Comment by dtzSiFive** [comment]
I edited the PR description to mention this, but for anyone following along via notifications:
You can see before/after differences from this change here: https://dtzsifive.github.io/circt-pp-results/ (running on FIRRTL's "regress" inputs presently).

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp
Comment (dtzSiFive) [reply lib/Conversion/ExportVerilog/ExportVerilog.cpp]: Thanks, fixed!

**Review by mikeurbach — APPROVED** [review]
Awww ExportVerilog is growing up so fast! Thanks for all the hard work on this.

This is a big change, and I don't think there's any sane way to do something other than a massive flip all at once. The fact that there weren't a ton of updates to existing tests seems like a good sign. Otherwise I think integration testing this on real designs and running simulation/lint is the best way to test it.

I have looked through all the changes to ExportVerilog, and nothing jumped out as weird. I'm not reviewing the actual output quality, since I think that's something we can easily tweak after the big change is in. Giving a tentative 👍 based on what I've seen, but would appreciate at least another approver.

**Review by uenoku — APPROVED** [review]
It looks great to me. This will drastically improve the verilog output quality. I think we can incrementally improve further after merging the PR. Thank you for the great work!

File: test/Conversion/ExportVerilog/pretty.mlir
Comment (uenoku) [reply test/Conversion/ExportVerilog/pretty.mlir:27]: Thanks, I think we can improve this kind of details later so I'm happy to keep as is.

**Comment by dtzSiFive** [comment]
Thanks all!
Rebased to resolve conflicts, NFCI since last reviews (modulo one case where adjusted end of box to be after closing punctuation).

Used more `scopedBox`-style emission as I think it makes it a lot clearer how boxes are nested and their lifetimes (easier to read + harder to get wrong, will always be closed; mechanical change replacing streaming a begin/end with `scopedBox`), and introduced two new named Tokens to simplify use of common groups ("neverbox" and "bbox2").

**Comment by dtzSiFive** [comment]
Thanks all!

**Pull request merged**
